File: graphic/drawPieChart.R
#!/usr/bin/env Rscript

suppressMessages(library('getopt'))

command =  matrix(c(
    "help",         "h",   0,  "logical",     "Show help information",
    "input",        "i",   1,  "character",   "STAR mapping stats matrix from getStarMapStats.py",
    "label",        "l",   1,  "character",   "name of label column",
    "output",       "o",   1,  "character",   "output pdf",
    "title",        "t",   1,  "character",   "title for pie chart",
    "value",        "v",   1,  "character",   "name of value column (count)"
), byrow=TRUE, ncol=5)

# function
ShowHelp <- function(object, param, reverse=FALSE, bool=FALSE) {
  if (reverse) {
    if (bool) {
      judge <- !isTRUE(object)
    }else{
      judge <- !is.null(object)
    }
  }else{
    if (bool) {
      judge <- isTRUE(object)
    }else{
      judge <- is.null(object)
    }
  }
  if (judge) {
    if (param != 'none') {
      cat(paste("None valid ", param, "!\n"))
    }
    cat(paste(getopt(command, usage = T),"\n"))
    q()
  }
}

LoadPacakge <- function(name) {
  suppressMessages(library(name, character.only = TRUE))
  cat(paste("Load package: ", name, ".\n"))
}

# parsing arguments
args <- getopt(command)

ShowHelp(args$help, 'none', TRUE)
ShowHelp(args$input, '-i|--input')
ShowHelp(args$label, '-l|--label')
ShowHelp(args$value, '-v|--value')
ShowHelp(args$output, '-o|--output')

# default values
if ( is.null(args$title) ) { args$title = "Pie Chart" }
if ( is.null(args$output) ) { args$output = './piechart.pdf' }

# load libraries
LoadPacakge('plotly')

## read countData from input matrix
countData <- read.table(args$input, row.names=NULL, header = TRUE, sep="\t")
#
## draw piechart

p <- plot_ly(countData,
  labels = as.formula(paste('~', args$label, sep="")), 
  values = as.formula(paste('~', args$value, sep="")),
  type = 'pie') %>%
  layout(title = args$title,
         xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
         yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE))

options(warn=-1)

# webshot::install_phantomjs()
export(p, file = args$output)

File: graphic/drawPieChart.py
#!/usr/bin/env python3
import os
import sys
import argparse
import pandas as pd
import plotly
import plotly.express as px
import time

parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument('-i', '--input', action='store', type=str, required=True,
                    help='input matrix file')
parser.add_argument('-l', '--label', action='store', type=str, required=True,
                    help='name of label column')
parser.add_argument('-o', '--output', action='store', type=str, required=True,
                    help='output pdf')
parser.add_argument('-t', '--title', action='store', type=str,
                    default="Pie Chart",
                    help='title of the pie chart')
parser.add_argument('-v', '--value', action='store', type=str, required=True,
                    help='name of value column (count)')

args = parser.parse_args()
if len(sys.argv[1:]) == 0:
    parser.print_help()
    parser.exit()

#plotly.io.orca.config.executable = '/home/kzhou/.conda/envs/py3/bin/orca'
plotly.io.orca.ensure_server()
time.sleep(20)
#
df = pd.read_csv(args.input, sep='\t')

fig = px.pie(df, values=args.value, names=args.label, title=args.title)

plotly.io.write_image(fig, args.output)

#fig.write_image(args.output)
